[GEP-21] provider-local single-stack ipv6 implementation

nschad opened the pull request:
**How to categorize this PR?**
/area networking
/kind api-change
/kind enhancement

**What this PR does / why we need it**:

This PR introduces the rest of the code necessary for completing #7051.
This PR forms a unit with https://github.com/gardener/vpn2/pull/45 and https://github.com/gardener/gardener-extension-networking-calico/pull/301, the IPv6 feature only works together with all 3 PRs.

**Which issue(s) this PR fixes**:
Part of #7051 

**Special notes for your reviewer**:

Testing is fairly straightforward. However you will need to select the new gardener-extension-networking-calico aswell as the new vpn2 implementation. We did this by patching the following file.
```yaml
# example/provider-local/garden/base/kustomization.yaml
patches:
- patch: |
    apiVersion: core.gardener.cloud/v1beta1
    kind: ControllerDeployment
    metadata:
      name: networking-calico
    type: helm
    providerConfig:
      values:
        image:
          # snapshot including https://github.com/gardener/gardener-extension-networking-calico/pull/301
          repository: europe-docker.pkg.dev/gardener-project/snapshots/gardener/extensions/networking-calico
          tag: v1.38.0-dev-696ad3e056970c3c8e117ac9e98801f9b2f470d5
``` 

As well as modifying the following file to fetch our modified vpn2.

```diff
# imagevector/images.yaml
  name: vpn-seed-server
- tag: "0.19.0"
+ # snapshot including https://github.com/gardener/vpn2/pull/45
+ tag: "0.20.0-dev"

  name: vpn-shoot-client
  sourceRepository: github.com/gardener/vpn2
- tag: "0.19.0"
+ # snapshot including https://github.com/gardener/vpn2/pull/45
+ tag: "0.20.0-dev-580f1f2bc496d0847387908a09cb5e80f0c4c695"
```
**Release note**:
<!--
Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|noteworthy|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```feature developer
Add full single-stack IPv6 support for gardener provider-local 
```

timebertt committed:
Configure `Network.spec.ipFamilies` based on `Shoot`

Commit:
Add example shoot for IPv6

timebertt committed:
Adapt NetworkPolicies

timebertt committed:
Configure apiserver-proxy to listen on IPv6

timebertt committed:
Explicitly enable IPv6 forwarding

timebertt committed:
Adapt kube-apiserver configuration

timebertt committed:
Add helper to get address bit length

timebertt committed:
Configure components to listen on IPv6

timebertt committed:
Configure VPN EnvoyFilter to match both IP families

Commit:
configure vpnseedserver to work with different IPFamilies

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
configure vpnshoot to work with different IPFamilies

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

timebertt committed:
Validate network disjointness with VPN IPv6 range

Also, make sure that both IPv4 and IPv6 contexts have the same test cases.

timebertt committed:
Disable logging for local IPv6 seeds

ScheererJ commented:
Hi @nschad,

is this pull request supposed to be complete or is it still considered a draft? Some commits have `WIP` in their title, which I understand as work-in-progress.

nschad commented:
> Hi @nschad,
> 
> is this pull request supposed to be complete or is it still considered a draft? Some commits have `WIP` in their title, which I understand as work-in-progress.

Yep it's complete (I can remove the WIP from the commit). There might be 2 or 3 TODO still in the code of which we would happily want your input.

rfranzke commented on example/provider-local/shoot-ipv6.yaml:
```diff
@@ -0,0 +1,45 @@
+    shoot.gardener.cloud/infrastructure-cleanup-wait-period-seconds: "0"
+    shoot.gardener.cloud/cloud-config-execution-max-delay-seconds: "0"
+spec:
+  seedName: local
```
Remove this field

rfranzke commented on example/provider-local/shoot-ipv6.yaml:
```diff
@@ -0,0 +1,45 @@
+    # TODO(scheererj): Drop this once v1.32 has been released and https://github.com/gardener/gardener-extension-networking-calico/pull/250 is available as release
+    providerConfig:
+      apiVersion: calico.networking.extensions.gardener.cloud/v1alpha1
+      kind: NetworkConfig
```
Remove this. Please reconcile this manifest again against the others on `master` branch

nschad replied:
We will retest this then using no NetworkConfig because we haven't done it yet and I'm unsure if it will work without

rfranzke commented on example/provider-local/shoot-ipv6.yaml:
```diff
@@ -0,0 +1,45 @@
+      maxSurge: 1
+      maxUnavailable: 0
+  kubernetes:
+    version: 1.26.0
```
1.28?

rfranzke commented on pkg/utils/net/ip_test.go:
```diff
@@ -0,0 +1,25 @@
+package net
```
license header

rfranzke commented on pkg/utils/net/net_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
2023

rfranzke commented on pkg/utils/net/net_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+	. "github.com/onsi/gomega"
+)
+
+func TestRetry(t *testing.T) {
```
```suggestion
func TestNet(t *testing.T) {
```

rfranzke commented on pkg/utils/net/net_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+
+func TestRetry(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Net Suite")
```
```suggestion
	RunSpecs(t, "Utils Net Suite")
```

rfranzke commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -291,6 +292,17 @@ func (r *Reconciler) reconcileNetworkPolicyAllowToPublicNetworks(ctx context.Con
+				}, {
+					IPBlock: &networkingv1.IPBlock{
+						CIDR: "::/0",
+						// TODO: exclude private blocks so that this actually allows only public networks
```
Assign a name to this TODO. Either address it now or explain what needs to be done

nschad replied:
It's about potentially excluding these networks https://en.wikipedia.org/wiki/Private_network#Private_IPv6_addresses.
Just like you did with IPv4 Networks above. It is more of a question.

timebertt replied:
We tried our best to resolve it here and asked the expert in https://github.com/gardener/gardener/pull/8574/commits/e38b2b67b271ab5ae02022fdfa682c9c56c636b0#r1357015066 :)

rfranzke commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -474,12 +487,13 @@ func (r *Reconciler) reconcileNetworkPolicyAllowToDNS(ctx context.Context, log l
 					// required for node local dns feature, allows egress traffic to node local dns cache
 					{
 						IPBlock: &networkingv1.IPBlock{
+							// TODO: support node-local-dns with IPv6 single-stack networking
```
same here

rfranzke commented on pkg/component/vpnseedserver/vpn_seed_server.go:
```diff
@@ -868,20 +881,34 @@ func (v *vpnSeedServer) emptyEnvoyFilter() *networkingv1alpha3.EnvoyFilter {
 	return &networkingv1alpha3.EnvoyFilter{ObjectMeta: metav1.ObjectMeta{Name: v.namespace + "-vpn", Namespace: v.istioNamespaceFunc()}}
 }
 
+//TODO: unused function?
```
?

nschad replied:
The below function is unused just an FYI

rfranzke commented on pkg/component/vpnseedserver/vpn_seed_server.go:
```diff
@@ -868,20 +881,34 @@ func (v *vpnSeedServer) emptyEnvoyFilter() *networkingv1alpha3.EnvoyFilter {
       socket_address:
         protocol: TCP
-        address: 0.0.0.0
+        address: ` + fmt.Sprintf("%q", listenAddress) + `
```
why not
```suggestion
        address: "` + listenAddress + `"
```
?

same below

nschad replied:
https://pkg.go.dev/fmt#pkg-overview

```
%q	a single-quoted character literal safely escaped with Go syntax.
```

rfranzke commented on pkg/component/vpnshoot/vpnshoot.go:
```diff
@@ -81,6 +82,8 @@ type ReversedVPNValues struct {
 	Endpoint string
 	// OpenVPNPort is the port for the ReversedVPN.
 	OpenVPNPort int32
+	// IPFamilies are the IPFamilies of the shoot
```
```suggestion
	// IPFamilies are the IPFamilies of the shoot.
```

rfranzke commented on pkg/component/vpnshoot/vpnshoot_test.go:
```diff
@@ -687,6 +692,26 @@ status: {}
+				Expect(runtime.DecodeInto(newCodec(), managedResourceSecret.Data["deployment__kube-system__vpn-shoot.yaml"], deployment)).To(Succeed())
+				Expect(deployment).To(DeepEqual(deploymentFor(secretNameCA, secretNameClient, secretNameTLSAuth, values.VPAEnabled)))
+			})
+
```
```suggestion
```

rfranzke commented on pkg/apis/core/validation/shoot.go:
```diff
@@ -511,6 +511,7 @@ func ValidateNodeCIDRMaskWithMaxPod(maxPod int32, nodeCIDRMaskSize int32, networ
 func ValidateTotalNodeCountWithPodCIDR(shoot *core.Shoot) field.ErrorList {
 	allErrs := field.ErrorList{}
 
+	// TODO: clarify why this is only called for create requests but not for updates
```
cc @shafeeqes (introduced with https://github.com/gardener/gardener/pull/5389).
From the PR description:

> The validation is applied only to new shoots, to not break the existing clusters.

nschad replied:
CC @timebertt 

I guess we can drop the question PR then?

timebertt replied:
Yes, then our confusion is clarified.

Commit:
pr review: address formatting, added missing license

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

timebertt committed:
Reconcile example manifest with IPv4 counterpart

timebertt committed:
Address review suggestions by rfranzke (2)

timebertt committed:
Prefactor: simplify handling of `allPrivateNetworkBlocks`

timebertt committed:
Properly handle IPv6 in network policy controller

timebertt commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -282,26 +323,7 @@ func (r *Reconciler) reconcileNetworkPolicyAllowToPublicNetworks(ctx context.Con
-						// }, r.SeedNetworks.BlockCIDRs...),
-					},
-				}},
+				To: append(peersV4, peersV6...),
```
@ScheererJ @axel7born @DockToFuture
We tried our best to implement the `allow-to-public-networks` policy for IPv6 but we are not sure about the details.
Can you take a look at this commit to see if our implementation makes sense?

JuliusSte replied:
Hi,
short reminder: We are not sure how to proceed here, could you please check our implementation? Would be highly appreciated! @ScheererJ @axel7born @DockToFuture

Thank you!

axel7born commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -71,6 +75,23 @@ type RuntimeNetworkConfig struct {
+	var peers = append([]string(nil), r.BlockCIDRs...)
+
+	if len(r.IPFamilies) == 1 && r.IPFamilies[0] == ipFamily {
+		peers = append(peers, r.Pods, r.Services)
```
Don't we need to block seed ranges as well?

nschad replied:
@axel7born I really don't know. Can somebody please check it?

axel7born replied:
I'm not sure, what the original intention of this networkpolicy was. But my question should not block this PR.

axel7born commented on test/e2e/gardener/common.go:
```diff
@@ -86,7 +86,8 @@ func DefaultShoot(name string) *gardencorev1beta1.Shoot {
 			Networking: &gardencorev1beta1.Networking{
-				Type: pointer.String("calico"),
+				Type:       pointer.String("calico"),
+				IPFamilies: []gardencorev1beta1.IPFamily{gardencorev1beta1.IPFamilyIPv4},
```
Why is this needed? Shouldn't that be the default?
Is it only needed for the update test?

nschad replied:
It actually fixxed the `pull-gardener-e2e-kind-ha-multi-zone` e2e test. For the upgrade test we have a fix coming up hopefully soon™️. But yes IPv4 is the default, I do like explicit more then implicit though.

If you know why this fixxed the previous mentioned e2e test, please let us know 😆 

Might have been just flaky.

nschad replied:
We will drop the commit for now...

MichaelEischer committed:
Allow setting the default IPFamilies for network extension objects

This special case is necessary to allow upgrades from an older Gardener
version to succeed. In older versions, networks.extensions.gardener.cloud
objects do not contain an IPFamilies field. Thus, when upgrading a
cluster, the IPFamilies field has to be set once.

It is only allowed to set IPv4 as family because existing shoots only
use IPv4.

ScheererJ commented on pkg/component/extensions/network/network_test.go:
```diff
@@ -53,6 +53,11 @@ var _ = Describe("#Network", func() {
+		networkPodV6IP       = "2002:db8:1::"
+		networkPodV6Mask     = 48
+		networkServiceV6IP   = "2002:db8:3::"
+		networkServiceV6Mask = 108
```
Please switch to the documentation range ([RFC-3849](https://datatracker.ietf.org/doc/html/rfc3849)):

```suggestion
		networkPodV6IP       = "2001:db8:1::"
		networkPodV6Mask     = 48
		networkServiceV6IP   = "2001:db8:3::"
		networkServiceV6Mask = 108
```

ScheererJ commented on example/provider-local/shoot-ipv6.yaml:
```diff
@@ -0,0 +1,45 @@
+    type: calico
+    pods: 2002:db8:1::/48
+#    nodes: 2002:db8:2::/64 -> it doesn't make sense to specify this, node IPs are given by the seed cluster's pod CIDR
+    services: 2002:db8:3::/108
```
Please switch to the documentation range (`2001:db8::/32`) instead of the `6to4` range (`2002::/16`). Feel free to consult https://en.wikipedia.org/wiki/Reserved_IP_addresses#IPv6 for details.

ScheererJ commented on pkg/utils/net/ip.go:
```diff
@@ -0,0 +1,28 @@
+		return 32
+	}
+	return ip.BitLen()
+}
```
Why do we use this rather obscure way of determining the address length? In other words: Why is an error equal to an IPv4 address?
I do not think that we are so far in our IPv6 adoption journey that we can deem usage of IPv4 to be equal to an error ;-) .

ScheererJ commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -474,12 +516,13 @@ func (r *Reconciler) reconcileNetworkPolicyAllowToDNS(ctx context.Context, log l
 					// required for node local dns feature, allows egress traffic to node local dns cache
 					{
 						IPBlock: &networkingv1.IPBlock{
+							// node local dns feature is only supported for shoots with IPv4 single-stack networking
 							CIDR: fmt.Sprintf("%s/32", nodelocaldnsconstants.IPVSAddress),
```
Why is node-local-dns only supported with IPv4?

nschad replied:
No technical reason, we considered it out-of-scope for this PR. Is this something that would be required?

ScheererJ replied:
`node-local-dns` is not enabled per default in the local setup. Therefore, you can indeed argue that it does not need to be included in this pull request.
Nevertheless, I would still like to have this tracked somewhere. This could be an additional point in https://github.com/gardener/gardener/issues/7051 or a separate issue. Otherwise, we will most likely forget it.

WDYT?

nschad replied:
100% Agree. Let's track it!

ScheererJ replied:
Will you add it to https://github.com/gardener/gardener/issues/7051 or do you prefer to track it in a separate issue?

nschad replied:
Added it to #7051

ScheererJ commented on pkg/component/apiserverproxy/apiserver_proxy_test.go:
```diff
@@ -576,21 +266,43 @@ subjects:
+				It("should deploy the managed resource successfully", func() {
+					test(false)
+				})
 			})
```
This results in strange configurations with IPv4 advertise address, but otherwise IPv6 settings. Could you please fix this so that the IPv6 configuration is correct?

ScheererJ commented on pkg/component/apiserverproxy/apiserver_proxy_test.go:
```diff
@@ -61,437 +62,124 @@ var _ = Describe("APIServerProxy", func() {
-              - name: envoy.filters.http.router
-                typed_config:
-                  "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+		serviceYAML = `apiVersion: v1
```
Why did you move the configmap and daemonset around and not just use a function like in other tests?

(There is no reason to change it back, but it made reading the diff unnecessarily complicated.)

ScheererJ commented on pkg/component/kubeapiserverexposure/kube_apiserver_sni.go:
```diff
@@ -124,16 +127,20 @@ func (s *sni) Deploy(ctx context.Context) error {
 		envoyFilterSpec bytes.Buffer
 	)
 
+	sniValues := s.valuesFunc()
```
Why do we need this? Is this not the same as `values` a couple of lines above?

ScheererJ commented on pkg/component/kubeapiserverexposure/kube_apiserver_sni.go:
```diff
@@ -56,9 +57,10 @@ func init() {
+	Hosts                       []string
+	APIServerProxy              *APIServerProxy
+	IstioIngressGateway         IstioIngressGateway
+	APIServerClusterIPPrefixLen int
```
Please remove `APIServerClusterIPPrefixLen` from `SNIValues`. It is only set during `Deploy()` as far as I can tell and there is no reason why it is in this values struct. A local variable in `Deploy()` should be enough.

ScheererJ commented on pkg/component/nodeproblemdetector/node_problem_detector.go:
```diff
@@ -226,7 +225,7 @@ func (c *nodeProblemDetector) computeResourcesData() (map[string][]byte, error)
 								Command: []string{
 									"/bin/sh",
 									"-c",
-									"exec /node-problem-detector --logtostderr --config.system-log-monitor=/config/kernel-monitor.json,/config/docker-monitor.json,/config/systemd-monitor.json .. --config.custom-plugin-monitor=/config/kernel-monitor-counter.json,/config/systemd-monitor-counter.json .. --config.system-stats-monitor=/config/system-stats-monitor.json --prometheus-address=" + daemonSetPrometheusAddress + " --pr ...
```
According to https://github.com/kubernetes/node-problem-detector/blob/34b265af347190b4dff320d77711bc92a5e8342d/cmd/options/options.go#L126 it looks like not setting this option will default to `127.0.0.1`. Is this what we want here?

nschad replied:
Technically not but we noticed that some applications work just fine with binding set "127.0.0.1" and some do not. That's why for some it's just removed and for some we have it set explicitly

ScheererJ replied:
Ok, it does not look like we use this endpoint, e.g. for metrics. Hence, this should be fine.

ScheererJ commented on pkg/component/vpnseedserver/vpn_seed_server.go:
```diff
@@ -18,7 +18,9 @@ import (
 	"path/filepath"
+	"strings"
 
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
```
Can you please move this import to the other gardener imports?

ScheererJ commented on pkg/component/vpnseedserver/vpn_seed_server_test.go:
```diff
@@ -18,6 +18,7 @@ import (
 	"context"
 	"fmt"
 
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
```
Please move the import to the gardener block.

ScheererJ commented on pkg/apis/core/v1beta1/constants/types_constants.go:
```diff
@@ -715,8 +715,10 @@ const (
-	// DefaultVPNRange is the default network range for the VPN between seed and shoot cluster.
+	// DefaultVPNRange is the default IPv4 network range for the VPN between seed and shoot cluster.
 	DefaultVPNRange = "192.168.123.0/24"
+	// DefaultVPNRangeV6 is the default IPv6 network range for the VPN between seed and shoot cluster.
+	DefaultVPNRangeV6 = "fd8f:6d53:b97a:1::/120"
```
While I can only applaud concerning the correct usage of unique local addresses (ULA) with regards to the random global identifier, I wonder if we could use something more readable, but that is probably only a nit.

I hope that we do not plan to use ULA anywhere as we would otherwise have the potential of overlap...

nschad replied:
Yeah we can use whatever here, it's just an example

nschad replied:
If you have any suggestion, let me know.

ScheererJ replied:
It is not just an example, but the value we will use for the VPN in an IPv6 environment going forward.
Nevertheless, we can use this for now (and probably the foreseeable future).

ScheererJ commented on example/gardener-local/gardenlet/values-ipv6.yaml:
```diff
@@ -1,4 +1,7 @@
 config:
+  # loki is unable to cope with IPv6, hence disable logging entirely
+  logging:
+    enabled: false
```
Is this really true or do we just take a shortcut here? Looking into the loki repository it seems like some people actually have loki running with IPv6.

nschad replied:
No technical reason, we considered it out-of-scope for this PR. Is this something that would be required?

nschad replied:
With this introduced: https://github.com/grafana/loki/pull/9184

We can test if IPv6 support is now available

ScheererJ replied:
From my perspective, this should be at least included in https://github.com/gardener/gardener/issues/7051. It can be a separate pull request, though.

ScheererJ replied:
Please decide if you want to add it here or to the issue for a later pull request.

nschad replied:
Also added to the ☂️  issue ❤️ 

Thank you for not increasing the size of PR any more 😮‍💨

ScheererJ commented on example/provider-local/shoot-ipv6.yaml:
```diff
@@ -7,21 +7,15 @@ metadata:
-      apiVersion: calico.networking.extensions.gardener.cloud/v1alpha1
-      kind: NetworkConfig
+    pods: 2002:db8:1::/48
+    services: 2002:db8:3::/108
```
Please use the documentation range instead of the `6to4` range:

```suggestion
    pods: 2001:db8:1::/48
    services: 2001:db8:3::/108
```

ScheererJ commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -71,6 +75,23 @@ type RuntimeNetworkConfig struct {
 
+// GetBlockedNetworkPeers returns a list of CIDRs to exclude from a NetworkPolicy IPBlock. ipFamily should match the
+// IP family of the IPBlock.CIDR value. The resulting list still needs to be filtered for subsets of IPBlock.CIDR.
+func (r RuntimeNetworkConfig) GetBlockedNetworkPeers(ipFamily gardencore.IPFamily) []string {
```
Does this need to be public or can it be private? It looks like it is only used in this file. Hence, you should lower-case it.

ScheererJ commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -71,6 +75,23 @@ type RuntimeNetworkConfig struct {
 	BlockCIDRs []string
 }
 
+// GetBlockedNetworkPeers returns a list of CIDRs to exclude from a NetworkPolicy IPBlock. ipFamily should match the
```
```suggestion
// GetBlockedNetworkPeers returns a list of CIDRs to exclude from a NetworkPolicy IPBlock. `ipFamily` should match the
```

ScheererJ commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -272,6 +293,26 @@ func (r *Reconciler) reconcileNetworkPolicyAllowToAPIServer(ctx context.Context,
+		// In IPv6 however, cluster networks might be "public" (e.g., if using prefix delegation from provider).
+		// As this NetworkPolicy should only allow communication with public networks *outside* the cluster,
+		// we exclude the cluster networks.
+		r.RuntimeNetworks.GetBlockedNetworkPeers(gardencore.IPFamilyIPv6)...,
```
I am not sure I understand the reasoning here. The scenario may not be that relevant as seed clusters do not open connections to shoot clusters, though.
In the IPv4 world, I can connect from seed to shoot using this network policy assuming that the shoot's load balancers are exposed in the internet. However, with IPv6 this would not work anymore. This is at best confusing.

Please explain.

ScheererJ replied:
After a round of offline discussions, I think that we are fine with the current approach. This means that public networks are everything except the clusters and private ranges. Private networks are only the private range without the clusters.

ScheererJ commented on pkg/controller/networkpolicy/reconciler.go:
```diff
@@ -356,19 +372,31 @@ func (r *Reconciler) reconcileNetworkPolicyAllowToPrivateNetworks(ctx context.Co
+			if gardencorev1beta1.IsIPv4SingleStack(shoot.Spec.Networking.IPFamilies) {
+				blockedNetworkPeersV4 = append(blockedNetworkPeersV4, shootNetworks...)
+			} else {
+				blockedNetworkPeersV6 = append(blockedNetworkPeersV4, shootNetworks...)
```
Similar as above, I am a bit puzzled about adding IPv6 cluster addresses here to the private network part. This code replicates the old IPv4 code, but the naming (private vs. public) does not match what we do with IPv6.

Maybe we need to discuss this in more detail.

ScheererJ replied:
After a round of offline discussions, I think that we are fine with the current approach. This means that public networks are everything except the clusters and private ranges. Private networks are only the private range without the clusters.

nschad replied:
But I think this is actually wrong, shouldn't this be? 🤔 
```diff
-				blockedNetworkPeersV6 = append(blockedNetworkPeersV4, shootNetworks...)
+				blockedNetworkPeersV6 = append(blockedNetworkPeersV6, shootNetworks...)
```

nschad replied:
But we do still have the sync on Friday so let's wait until then.

ScheererJ replied:
Good catch. With all the `v4` and `v6` I overlooked it.
I agree that the `else` block should not have IPv4 at all in it. (At least, as long as we only consider only single-stack deployments.)

nschad replied:
Should be fixed now. I just personally don't know if we configured this right

ScheererJ reviewed (changes requested):
Thanks a lot for the changes and helping Gardener get ready for IPv6. I have a few comments and questions.

Commit:
change ipv6 range prefix from 6to4 to use doc range (2001:db8::/32)

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
make getBlockedNetworkPeers private

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
remove duplicate `s.valuesFunc()` in kube_apiserver_sni.go

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
set correct advertiseIPAddress for IPv6 test

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
remove `APIServerClusterIPPrefixLen` from sniValues

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
rework `helper.GetBitLen()`

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

Commit:
fix `blockedNetworkPeersV6` to correctly use IPv6 addresses instead of IPv4

Signed-off-by: Niclas Schad <niclas.schad@stackit.de>

DockToFuture commented on pkg/component/vpnseedserver/vpn_seed_server.go:
```diff
@@ -870,13 +883,26 @@ func getLabels() map[string]string {
+		listenAddress = "0.0.0.0"
+		// we don't want to use AUTO for single-stack clusters as it causes an unnecessary failed lookup
+		// ref https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#enum-config-cluster-v3-cluster-dnslookupfamily
+		dnsLookupFamily = "V4_ONLY"
```
```suggestion
		dnsLookupFamily = "V4_PREFERRED"
```

DockToFuture commented on pkg/component/vpnseedserver/vpn_seed_server.go:
```diff
@@ -870,13 +883,26 @@ func getLabels() map[string]string {
       socket_address:
         protocol: TCP
-        address: 0.0.0.0
+        address: "` + listenAddress + `"
```
add an additional listener
```
    additional_addresses:
    - address:
        socket_address:
          address: "` + listenAddressV6 + `"
          port_value: ` + fmt.Sprintf("%d", EnvoyPort) + `
```

DockToFuture reviewed (changes requested):
Please adapt the envoy configuration to also allow IPv4 Seeds with IPv6 Shoots.

DockToFuture commented on pkg/component/vpnseedserver/vpn_seed_server.go:
```diff
@@ -870,13 +883,26 @@ func getLabels() map[string]string {
+		dnsLookupFamily = "V4_ONLY"
+	)
+
+	if gardencorev1beta1.IsIPv6SingleStack(v.values.Network.IPFamilies) {
```
Replace the 4 following lines with `listenAddressV6 := "::"`

ScheererJ reviewed (approved):
Thanks a lot for the changes.

/lgtm
/approve

DockToFuture reviewed (approved):
Then lets address this in another PR once the change is in.
/lgtm

timebertt reviewed (commented):
Thanks @nschad for rebasing.
Let's get this in before the next conflicts arise :)

/hold cancel
/lgtm
/approve
